webSockets/sender/sender.go

```diff
@@ -0,0 +1,181 @@
+package sender
+
+import (
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/multiversx/mx-chain-core-go/core"
+	"github.com/multiversx/mx-chain-core-go/core/check"
+	"github.com/multiversx/mx-chain-core-go/core/closing"
+	"github.com/multiversx/mx-chain-core-go/webSockets"
+	"github.com/multiversx/mx-chain-core-go/webSockets/connection"
+	outportData "github.com/multiversx/mx-chain-core-go/webSockets/data"
+)
+
+// ArgsSender holds the arguments needed for creating a web-sockets sender
+type ArgsSender struct {
+	WithAcknowledge        bool
+	RetryDurationInSeconds int
+	PayloadConverter       webSockets.PayloadConverter
+	Log                    core.Logger
+}
+
+type sender struct {
+	counter         uint64
+	withAcknowledge bool
+	connections     ConnectionsHandler
+	payloadParser   webSockets.PayloadConverter
+	log             core.Logger
+	safeCloser      core.SafeCloser
+	retryDuration   time.Duration
+}
+
+func NewSender(args ArgsSender) (*sender, error) {
+	err := checkArgs(args)
+	if err != nil {
+		return nil, err
+	}
+
+	return &sender{
+		counter:         0,
+		safeCloser:      closing.NewSafeChanCloser(),
+		connections:     connection.NewWebsocketClientsHolder(),
+		log:             args.Log,
+		payloadParser:   args.PayloadConverter,
+		withAcknowledge: args.WithAcknowledge,
+		retryDuration:   time.Duration(args.RetryDurationInSeconds) * time.Second,
+	}, nil
+}
+
+// AddConnection will add in the connections map the provided connection
+func (s *sender) AddConnection(client webSockets.WSConClient) error {
+	return s.connections.AddClient(client)
+}
+
+// Send will prepare and send the provided WsSendArgs
+func (s *sender) Send(args outportData.WsSendArgs) error {
+	assignedCounter := atomic.AddUint64(&s.counter, 1)
+	newPayload := s.payloadParser.ConstructPayloadData(args, assignedCounter, s.withAcknowledge)
+
+	return s.send(newPayload, assignedCounter)
+}
+
+func (s *sender) send(payload []byte, assignedCounter uint64) error {
+	clients := s.connections.GetAll()
+	if len(clients) == 0 {
+		return outportData.ErrNoClientToSendTo
```

Comment:
is this an error? maybe I want to leave the server open forever and from time to time a client connects to it. will this work?

Reply:
Not sure if this is the behavior that we want. 
In case the node is a server and there are no clients that are connected I want the node to remain stuck until a client will connect.